### ISSUE TITLE
Validate US Telephone Numbers Accepts Invalid Characters

### ISSUE THREAD

**Issue by CharmedSatyr** [FreeCodeCamp/FreeCodeCamp#8256]
#### Challenge Name

Validate US Telephone Numbers
https://www.freecodecamp.com/challenges/validate-us-telephone-numbers

#### Issue Description

It's possible to pass telephone numbers that include 10 numbers interspersed with random, invalid characters, or 11 numbers that begin with 1, interspersed with random, invalid characters, as ```true```, so long as the ```(``` and ```)```, for example, are validly spaced.

For example, my code, before test 5, passes ``` telephoneCheck("(555)abcdefghijklmnop55*5?-5/5?#5qrstuvwxyz5"); ``` as ```true``` and also passes the challenge.

#### Your Code
```js
function telephoneCheck(str) {
  
  var num = /[\d]/;
  var justDigits = [];
  var bool = true;
  
  //the str array retains all characters from the original str
  str = str.split('');
  
  //push non-string numbers to array justDigits
  for (var x in str) {
    if (str[x].match(num)) {
      justDigits.push(parseFloat(str[x]));
    }
  }

  //test 1 - If justDigits has 10 numbers or 1 + 10 numbers, it'll probably be OK. Otherwise the phone # is invalid.
  if (justDigits.length === 10 || justDigits.length === 11 && justDigits[0] === 1) {} else {
    bool = false;
  }

  //test 2 - str should start with a number or a '('
  if (isNaN(str[0]) === false || (str[0] === '(')) {} else {
    bool = false;
  }

  //test 3 - if a str has a '(', the ')' should be 4 spots later
  if (str.indexOf('(') > -1) {
    if (str.indexOf(')') === str.indexOf('(') + 4) {} else {
      bool = false;
    }
  }

  //test 4 - if a str has a ')', the '(' should be 4 spots before and should exist
  if (str.indexOf(')') > -1) {
    if (str.indexOf('(') === str.indexOf(')') - 4 && str.indexOf('(') > -1) {} else {
      bool = false;
    }
  }
/*
  //test 5 - not required to pass, but a good additional test would check for invalid characters
  var invalid = /[^\d()-\s]/;
  for (var y in str) {
    if (str[y].match(invalid)) {
      bool = false;
    }
  }
*/
  return bool;
}

telephoneCheck("(555)abcdefghijklm(nop)55*5?-5/5?#5qrstuvwxyz5"); //Currently passes as true
```

It would be helpful to require some variant of the following (my test 5):
```js
  var invalid = /[^\d()-\s]/;
  for (var y in str) {
    if (str[y].match(invalid)) {
      bool = false;
    }
  }
```

**Comment by khaduch** [comment]
If you write your code to not allow characters other than numbers, then it would not allow characters other than numbers.  It appears that you are counting numeric characters and accepting that there is a certain number of numeric characters, regardless of which other characters are also within the string.

If you write a regular expression that will only match characters, for example `/\d{3}/` to match the three-digit portion of the string, then it will not match anything like "3e3e3e" and accept that as a valid phone number (in this case, just a portion of the phone number.)

Basically, you can write a one-line regular expression that will allow only numbers and the potential spaces, dashes, parentheses, and nothing else.

**Comment by CharmedSatyr** [comment]
You are right. There are many ways to fix the problem, and I have passed the test. However, my point is that very bad numbers pass as true with my current code (excluding test 5). I'm pointing out that FCC should update its tests to to fail my example.

**Comment by raisedadead** [comment]
Agreed. Thanks for reporting this @CharmedSatyr

**Comment by khaduch** [comment]
Does your code pass all of the tests in the test suite for that challenge?

If I run your example through my validation function, it returns `false` as the answer, because my regexp is not allowing the sequence of characters that you have in your test?  Perhaps I'm just missing something? I just think that you have incorrect code for checking this?

Or are you saying that a test should be provided that should should an incorrect validation of a number with code as you have written?  Perhaps that is the point that I'm missing...

**Comment by CharmedSatyr** [comment]
@khaduch Yep, it does pass the current test suite. Without test 5, my example is true, according to the test suite. There should be a test that strengthens the requirement for regexp to use something like in my test 5 or like you are using.

@raisedadead I should also mention that my test passes numbers as true if they have their '(' and ')' properly spaced, even if the parentheses are in the wrong spots. For example, ```telephoneCheck("(555)5(55?)-5555");``` currently passes as ```true```.

**Comment by erictleung** [comment]
Closing as #8900 resolves this issue. Happy coding all!
